src/query/pipeline/sinks/src/async_mpsc_sink.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use async_trait::unboxed_simple;
+use common_exception::Result;
+use common_expression::DataBlock;
+use common_pipeline_core::processors::port::InputPort;
+use common_pipeline_core::processors::processor::Event;
+use common_pipeline_core::processors::Processor;
+
+/// Sink with multiple inputs.
+#[async_trait]
+pub trait AsyncMpscSink: Send {
+    const NAME: &'static str;
+
+    async fn on_start(&mut self) -> Result<()> {
+        Ok(())
+    }
+
+    async fn on_finish(&mut self) -> Result<()> {
+        Ok(())
+    }
+
+    #[unboxed_simple]
+    async fn consume(&mut self, data_block: Vec<DataBlock>) -> Result<bool>;
+}
+
+pub struct AsyncMpscSinker<T: AsyncMpscSink + 'static> {
+    inner: T,
+    finished: bool,
+    inputs: Vec<Arc<InputPort>>,
+    inputs_data: Option<Vec<DataBlock>>,
```

Comment:
```suggestion
    inputs_data: Option<DataBlock>,
```

Reply:
Why not consume a batch of `DataBlock`s one time?

Reply:
We only need to consume all inputs once in first event.